jooq/src/main/java/io/micronaut/configuration/jooq/JooqConfigurationFactory.java

```diff
@@ -0,0 +1,53 @@
+package io.micronaut.configuration.jooq;
+
+import io.micronaut.context.ApplicationContext;
+import io.micronaut.context.annotation.EachBean;
+import io.micronaut.context.annotation.Factory;
+import io.micronaut.context.annotation.Parameter;
+import io.micronaut.inject.qualifiers.Qualifiers;
+import org.jooq.*;
+import org.jooq.conf.Settings;
+import org.jooq.impl.DefaultConfiguration;
+import org.jooq.impl.DefaultDSLContext;
+
+import javax.sql.DataSource;
+
+@Factory
+public class JooqConfigurationFactory {
+
+    @EachBean(DataSource.class)
+    public DefaultConfiguration jooqConfiguration(
+            @Parameter String name,
+            DataSource dataSource,
+            ApplicationContext ctx
+    ) {
+        DefaultConfiguration configuration = new DefaultConfiguration();
+
+        JooqConfigurationProperties properties = ctx.findBean(JooqConfigurationProperties.class, Qualifiers.byName(name))
+                .orElseGet(JooqConfigurationProperties::new);
+        configuration.setSQLDialect(properties.determineSqlDialect(dataSource));
+
+        configuration.setDataSource(dataSource);
+        ctx.findBean(TransactionProvider.class, Qualifiers.byName(name)).ifPresent(configuration::setTransactionProvider);
+        ctx.findBean(Settings.class, Qualifiers.byName(name)).ifPresent(configuration::setSettings);
+        ctx.findBean(ExecutorProvider.class, Qualifiers.byName(name)).ifPresent(configuration::setExecutorProvider);
+        ctx.findBean(RecordMapperProvider.class, Qualifiers.byName(name)).ifPresent(configuration::setRecordMapperProvider);
+        ctx.findBean(RecordUnmapperProvider.class, Qualifiers.byName(name)).ifPresent(configuration::setRecordUnmapperProvider);
+        configuration.setExecuteListenerProvider(ctx.getBeansOfType(ExecuteListenerProvider.class, Qualifiers.byName(name))
+                .toArray(new ExecuteListenerProvider[0]));
+        configuration.setRecordListenerProvider(ctx.getBeansOfType(RecordListenerProvider.class, Qualifiers.byName(name))
+                .toArray(new RecordListenerProvider[0]));
+        configuration.setVisitListenerProvider(ctx.getBeansOfType(VisitListenerProvider.class, Qualifiers.byName(name))
+                .toArray(new VisitListenerProvider[0]));
+        configuration.setTransactionListenerProvider(ctx.getBeansOfType(TransactionListenerProvider.class, Qualifiers.byName(name))
+                .toArray(new TransactionListenerProvider[0]));
```

Comment:
While these `XYZProvider` types do allow for adding an extra level of abstraction and indirection, most people probably use `RecordListener` directly, instead of `RecordListenerProvider`. Without knowing much about micronaut's DI mechanisms, is there a micronaut-specific reason not to support the actual SPIs here, but their providers?

Also, possibly missing SPIs:

- `Clock` (great for some unit/integration test cases)
- `MetaProvider` (this will become increasingly important from jOOQ 3.12 onwards)
- `DiagnosticsListener` (possibly, not sure about this one)

Reply:
Do you have any idea why Spring Boot integration does not allow injecting non-provider beans? I just made it in a similar way. My guess is that mixing both provider-based DI and direct SPI injection can be error-prone in some cases.

Reply:
> Do you have any idea why Spring Boot integration does not allow injecting non-provider beans?

I don't, nor was I aware of this in Spring Boot

> My guess is that mixing both provider-based DI and direct SPI injection can be error-prone in some cases.

Well, they override each other, so that's a bit of a "danger", of course.

Reply:
I think I'll stick to provider injection to keep it simple. With Micronaut DI it is really easy for end-user to automatically create e.g. `RecordListenerProvider` bean for every `RecordListener` with just a few lines of code.
If we do it here, there will be no easy way to disable which I imagine can be needed sometimes to avoid injecting some extension beans twice (as a direct bean and as a provider bean).